src/lib/config/software.js

```diff
@@ -48,3 +67,254 @@ export const appQueryFragments = `fragment Software on AppEnvironmentSoftwareSet
+			throw new UserError( `Component ${ userProvidedComponent } is not supported. Use one of: ${ validComponents.join( ',' ) }` );
+		}
+		return userProvidedComponent;
+	}
+
+	if ( validComponents.length === 0 ) {
+		throw new UserError( 'No components are supported for this application' );
+	}
+
+	if ( validComponents.length === 1 ) {
+		return validComponents[ 0 ];
+	}
+
+	const choices = validComponents.map( item => ( {
+		message: COMPONENT_NAMES[ item ],
+		value: item,
+	} ) );
+	const select = new Select( {
+		message: 'Component to update',
+		choices,
+	} );
+	return await select.run();
+};
+
+const _processComponentVersion = async ( softwareSettings, component: string, userProvidedVersion: string | undefined ) => {
+	const versionChoices = _optionsForVersion( softwareSettings[ component ].options, softwareSettings[ component ].current.version );
+
+	if ( userProvidedVersion ) {
+		const validValues = versionChoices.map( item => item.value );
+		if ( ! validValues.includes( userProvidedVersion ) ) {
+			throw new UserError( `Version ${ userProvidedVersion } is not supported for ${ COMPONENT_NAMES[ component ] }. Use one of: ${ validValues.join( ',' ) }` );
+		}
+		return userProvidedVersion;
+	}
+
+	const versionSelect = new Select( {
+		message: `Version for ${ COMPONENT_NAMES[ component ] } to upgrade to`,
+		choices: versionChoices,
+	} );
+	return await versionSelect.run();
+};
+
+interface UpdateData {
+	component: string,
+	version: string,
+}
+
+export interface UpdatePromptOptions {
+	component?: string,
+	version?: string,
+	force?: boolean,
+}
+
+export const promptForUpdate = async ( appTypeId: number, opts: UpdatePromptOptions, softwareSettings ): UpdateData => {
+	const component = await _processComponent( appTypeId, opts.component );
+	const version = await _processComponentVersion( softwareSettings, component, opts.version );
+
+	const confirm = opts.force || await new Confirm( {
+		message: `Are you sure you want to upgrade ${ COMPONENT_NAMES[ component ] } to ${ version }?`,
+	} ).run();
+
+	if ( confirm ) {
+		return {
+			component,
+			version,
+		};
+	}
+
+	throw new UserError( 'Update canceled' );
+};
+
+interface TrigerUpdateOptions {
+	appId: number,
+	envId: number,
+	component: string,
+	version: string,
+}
+
+export const triggerUpdate = async ( variables: TrigerUpdateOptions ) => {
+	debug( 'Triggering update', variables );
+	const api = await API();
+
+	return await api.mutate( { mutation: updateSoftwareMutation, variables } );
+};
+
+const _getLatestJob = async ( appId: number, envId: number ) => {
+	const api = await API();
+	let latestJob = null;
+	const result = await api.query( { query: updateJobQuery, variables: { appId, envId }, fetchPolicy: 'network-only' } );
+	const jobs = result?.data?.app?.environments[ 0 ].jobs || [];
+	for ( const job of jobs ) {
+		if ( latestJob ) {
+			if ( job.createdAt > latestJob.createdAt ) {
+				latestJob = job;
+			}
+		} else {
+			latestJob = job;
+		}
+	}
```

Comment:
This feels a bit simpler:

```
const jobs = [ { createdAt: 3 }, { createdAt: 1 }, { createdAt: 4 }, { createdAt: 2  } ];
const latestJob = jobs.reduce( ( prev, current ) => ( prev.createdAt > current.createdAt ) ? prev : current );
```

Reply:
I changed it. It needed the empty array check, but still probably simpler.